pkg/azurefile/controllerserver.go

```diff
@@ -986,6 +989,120 @@ func (d *Driver) ListSnapshots(_ context.Context, _ *csi.ListSnapshotsRequest) (
 	return nil, status.Error(codes.Unimplemented, "")
 }
 
+// restoreSnapshot restores from a snapshot
+func (d *Driver) restoreSnapshot(ctx context.Context, req *csi.CreateVolumeRequest, dstAccountName, dstAccountSasToken string, authAzcopyEnv []string, secretName, secretNamespace string, secrets map[string]string, shareOptions *fileclient.ShareOptions, accountOptions *azure.AccountOptions, storageEndpointSuffix string) error {
+	if shareOptions.Protocol == storage.EnabledProtocolsNFS {
+		return fmt.Errorf("protocol nfs is not supported for snapshot restore")
+	}
+	var sourceSnapshotID string
+	if req.GetVolumeContentSource() != nil && req.GetVolumeContentSource().GetSnapshot() != nil {
+		sourceSnapshotID = req.GetVolumeContentSource().GetSnapshot().GetSnapshotId()
+	}
+	resourceGroupName, srcAccountName, srcFileShareName, _, _, _, err := GetFileShareInfo(sourceSnapshotID) //nolint:dogsled
+	if err != nil {
+		return status.Error(codes.NotFound, err.Error())
+	}
+	snapshot, err := getSnapshot(sourceSnapshotID)
+	if err != nil {
+		return status.Error(codes.NotFound, err.Error())
+	}
+	dstFileShareName := shareOptions.Name
+	if srcFileShareName == "" || dstFileShareName == "" {
+		return fmt.Errorf("srcFileShareName(%s) or dstFileShareName(%s) is empty", srcFileShareName, dstFileShareName)
+	}
+	srcAccountSasToken := dstAccountSasToken
+	if srcAccountName != dstAccountName {
+		var srcAccountKey string
+		if srcAccountKey, err = d.cloud.GetStorageAccesskey(ctx, d.cloud.SubscriptionID, srcAccountName, resourceGroupName, true); err != nil {
+			return fmt.Errorf("get src accountkey get err: %v", err)
+		}
+		klog.V(2).Infof("generate sas token for resource group(%s) src account(%s)", resourceGroupName, srcAccountName)
+		if srcAccountSasToken, err = d.generateSASToken(srcAccountName, srcAccountKey, storageEndpointSuffix, d.sasTokenExpirationMinutes); err != nil {
+			return err
+		}
+		if dstAccountSasToken == "" {
```

Comment:
It's because if `srcAccountName != dstAccountName`, it means use different storage class to restore. Then in this case we should only use sas token, and then we should always generate srcAccountSasToken in L1020 and if dstAccountSasToken is empty we would generate it in L1025. And if ``srcAccountName == dstAccountName`, only `srcAccountSasToken := dstAccountSasToken` in L1013 is ok

Reply:
maneged identity auth does not support cross account copy?

Reply:
test it, managed identity auth also support cross account copy, it's ignored before. Only when `dstAccountSasToken != ""` do it